pageserver/src/tenant/layer_map.rs

```diff
@@ -277,30 +282,70 @@ where
     /// Helper function for BatchedUpdates::remove_historic
     ///
     pub fn remove_historic_noflush(&mut self, layer: Arc<L>) {
-        let kr = layer.get_key_range();
-        let lr = layer.get_lsn_range();
-        self.historic.remove(historic_layer_coverage::LayerKey {
-            key: kr.start.to_i128()..kr.end.to_i128(),
-            lsn: lr.start.0..lr.end.0,
-            is_image: !layer.is_incremental(),
-        });
+        self.historic
+            .remove(historic_layer_coverage::LayerKey::from(&*layer));
 
         if Self::is_l0(&layer) {
             let len_before = self.l0_delta_layers.len();
-
-            // FIXME: ptr_eq might fail to return true for 'dyn'
-            // references.  Clippy complains about this. In practice it
-            // seems to work, the assertion below would be triggered
-            // otherwise but this ought to be fixed.
-            #[allow(clippy::vtable_address_comparisons)]
             self.l0_delta_layers
-                .retain(|other| !Arc::ptr_eq(other, &layer));
-            assert_eq!(self.l0_delta_layers.len(), len_before - 1);
+                .retain(|other| !Self::compare_arced_layers(other, &layer));
+            // this assertion is related to use of Arc::ptr_eq in Self::compare_arced_layers,
+            // there's a chance that the comparison fails at runtime due to it comparing (pointer,
+            // vtable) pairs.
+            assert_eq!(
+                self.l0_delta_layers.len(),
+                len_before - 1,
+                "failed to locate removed historic layer from l0_delta_layers"
+            );
         }
 
         NUM_ONDISK_LAYERS.dec();
     }
 
+    /// Replace existing layer with another on having downloaded or layer eviction.
+    pub(self) fn replace_historic_noflush(
+        &mut self,
+        expected: &Arc<L>,
+        new: Arc<L>,
+    ) -> anyhow::Result<bool> {
+        let key = historic_layer_coverage::LayerKey::from(&**expected);
+        let other = historic_layer_coverage::LayerKey::from(&*new);
+
+        let was_l0 = Self::is_l0(expected);
+
+        anyhow::ensure!(
+            was_l0 == Self::is_l0(&new),
+            "replaced must both be l0 deltas or neither"
+        );
+        anyhow::ensure!(key == other, "replaced must have same layerkeys");
+
+        let l0_index = if was_l0 {
```

Comment:
We don't have test coverage for this `if was_l0 { ... } else { ... }` branch, do we?

Reply:
I didn't look at coverage reports. Are you thinking it might overwrite a wrong index?

Reply:
I'll see this CI run and put a todo!() there.

Reply:
Just making sure before I do the todo!() run -- there's an error for the case of l0 delta layer not being found from the layers, this is similar to the `assert_eq!(before, l0_deltas.len() - 1)`, but I just check it as a pre condition and not use an assert.

I think the `else` branch is not interesting, because nothing should be done for non-l0 delta layers.

Reply:
I was using coverage in the broad sense here. All tests added by this PR are in `historic_layer_coverage.rs`.
So, the aspect of "need to remember to update `l0_delta_layers`" isn't covered by unit tests.

Which is probably fine, the other methods don't have such coverage either.

Feel free to mark this as resolved.

Reply:
Aah yes, failed to think about unit tests because "pageserver code", but we can and should write some for the layermap. I'll add some for the add/replace/remove for non-l0 and l0.

Reply:
Tests: [c4f4dfd](https://github.com/neondatabase/neon/pull/3513/commits/c4f4dfd82e582c89ebf722ae6bc3ff02be88375f) followed by some refactoring-with-tests :)